text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,115 @@
+
+### ImmutableData XOR-URL
+XOR-URLs for ImmutableData’s are the simplest ones since they don’t need any additional information to uniquely identify them on the network, as opposed to MutableData’s that also have a type tag. Therefore an ImmutableData XOR-URL can be simply defined as `safe://<encoded ImmutableData XOR addr>`.
+
+### MutableData XOR-URL
+As already mentioned above, a string based on the XOR address along with a type tag is needed to uniquely identify a MutableData on the network, therefore the XOR-URL for a MD needs to include the type tag in it.
+
+When a MutableData is fetched, if it’s not an NFS/Files container with an `index.html` file, the resolver (e.g. the `webFetch` function) can return the MutableData’s raw data so the browser (or any client application) can render it in a specific way (see below for more details of what’s being proposed here in this regard).
+
+### MutableData versions
+MutableData’s are versioned and therefore this shall be also accountable in the XOR-URL format to allow any MD URL to (optionally) reference a specific version. This will be also necessary in the future when append-only type of data is made available on the network.
+
+The version value can be used to enforce a specific version to be retrieved, and otherwise fail if that specific version is not found. On the other hand, the latest version will be retrieved if the version value is omitted from the XOR-URL.
+
+### Paths
+Given that a referenced MutableData can effectively be the root NFS/Files container of the files of a safesite, any MD XOR-URL can also specify a path which needs to be resolved, just like when using the DNS/public-names system to resolve the path in a URL.
+
+### Browsable content
+Currently, when a public-name URL is resolved to an NFS/Files container which doesn’t have an `index.html` file, the browser simply shows an error stating that the safesite content was not found.
+
+For a MD XOR-URL that doesn’t have a path and there is no `index.html` file, the resolver can return the raw content of the MD (i.e. its key-value entries), and the browser can automatically generate an HTML page which makes the content browsable, generating links to other data when an entry’s key/value is a `safe://` string, in an analogous/similar way to how web servers on the current internet allow browsing on folders/directories.
+
+An specific MD entry key could be also supported (e.g. `__non_browsable`), that the owner of the MD can insert in the MD if the “browsable” feature should not be enabled. Although this cannot be really enforced but offered as a feature to be optionally supported by some clients like our SAFE browser.
```

Comment:
Would a client support `__non_browsable` but not support displaying the data in some format?

Reply:
I hesitated about having this part in this RFC since it's not really something related, and it in our case specific to a feature we can have in the browser I think. As you say, the client app can do anything with it really, so it's more about a convention and having our browser to follow it, but this cannot be enforced. In any case I'd agree to perhaps remove this from here and discuss it separately as a potential feature for the browser, and/or for our resolvable maps?

Reply:
> I'd agree to perhaps remove this from here and discuss it separately as a potential feature for the browser, and/or for our resolvable maps?

Yeh I think it's more browser based. We can discuss in an issue or so. I don't think it's resolvable maps related as it's not really a feature /requirement of the of the data structs etc. (saying they MUST display the file contents when no resolution is found seems over prescriptive).

Reply:
agreed, I'll remove that part then